### PR TITLE
feat(applications): add pagination middleware to findAllApplications

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Add pagination middleware to findAllApplications

1. update the `findAllApplications` query method to take limit and offset as its query input
2. add `koa-pagination` middle to the GET /applications api
3. Return total num of applications as totalCount to the pagination middleware

## Testing

Test locally with post man

Request:
image

Response:
image

Response Header:
image

@logto-io/eng

### CODE REVIEW

File: packages/core/src/queries/utils.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { sql, IdentifierSqlTokenType } from 'slonik';
+
+import pool from '@/database/pool';
+
+export const totalRowCount = async (table: IdentifierSqlTokenType) =>
+  pool.one<{ count: number }>(sql`
+    select count(*)
```
Comment (wangsijie) [inline packages/core/src/queries/utils.ts:7]: nit: `select count(*) as count`, but maybe slonik already did this.

**Review by wangsijie — APPROVED** [review]
LGTM

File: packages/core/src/queries/utils.ts (hunk 1 above)
Comment (simeng-li) [reply packages/core/src/queries/utils.ts:7]: yep slonik did do the part. the returned data type is { count: 4 }

File: packages/core/src/queries/application.ts [hunk 2]
```diff
@@ -7,12 +7,18 @@ import { buildUpdateWhere } from '@/database/update-where';
...
 import RequestError from '@/errors/RequestError';
 
+import { totalRowCount } from './utils';
+
 const { table, fields } = convertToIdentifiers(Applications);
 
-export const findAllApplications = async () =>
+export const findTotalNumberOfApplications = async () => totalRowCount(table);
+
+export const findAllApplications = async (limit: number, offset: number) =>
```
Comment (wangsijie) [inline packages/core/src/queries/application.ts:16]: try `buildFindMany`?

**Review by wangsijie — CHANGES_REQUESTED** [review]

File: packages/core/src/queries/application.ts (hunk 2 above)
Comment (simeng-li) [reply packages/core/src/queries/application.ts:19]: updated Please check

**Review by wangsijie — APPROVED** [review]
LGTM

File: packages/core/src/queries/application.ts [hunk 3]
```diff
@@ -1,19 +1,23 @@
...
 import { sql } from 'slonik';
 
+import { buildFindMany } from '@/database/find-many';
 import { buildInsertInto } from '@/database/insert-into';
 import pool from '@/database/pool';
 import { buildUpdateWhere } from '@/database/update-where';
 import { convertToIdentifiers, OmitAutoSetFields } from '@/database/utils';
 import RequestError from '@/errors/RequestError';
 
+import { totalRowCount } from './utils';
```
Comment (gao-sun) [inline packages/core/src/queries/application.ts:11]: will it be better to put in `@/database`?

File: packages/core/src/queries/utils.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { sql, IdentifierSqlTokenType } from 'slonik';
+
+import pool from '@/database/pool';
+
+export const totalRowCount = async (table: IdentifierSqlTokenType) =>
```
Comment (gao-sun) [inline packages/core/src/queries/utils.ts:5]: will it be better to start with verb, say `getTotalRowCount`?

File: packages/core/src/routes/application.ts [hunk 5]
```diff
@@ -17,8 +19,15 @@ import { AuthedRouter } from './types';
 const applicationId = buildIdGenerator(21);
 
 export default function applicationRoutes<T extends AuthedRouter>(router: T) {
-  router.get('/applications', async (ctx, next) => {
-    ctx.body = await findAllApplications();
+  router.get('/applications', koaPagination(), async (ctx, next) => {
+    const { limit, offset } = ctx.pagination;
+    const { count } = await findTotalNumberOfApplications();
```
Comment (gao-sun) [inline packages/core/src/routes/application.ts:24]: shall we use `Promise.all` to accelerate?

**Comment by simeng-li** [comment]
@gao-sun updated please double-check thx

**Pull request merged**